{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ReadTimeout",
     "evalue": "HTTPSConnectionPool(host='api.polygon.io', port=443): Read timed out. (read timeout=10)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTimeoutError\u001b[0m                              Traceback (most recent call last)",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/urllib3/connectionpool.py:449\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    445\u001b[0m         \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    446\u001b[0m             \u001b[39m# Remove the TypeError from the exception chain in\u001b[39;00m\n\u001b[1;32m    447\u001b[0m             \u001b[39m# Python 3 (including for exceptions like SystemExit).\u001b[39;00m\n\u001b[1;32m    448\u001b[0m             \u001b[39m# Otherwise it looks like a bug in the code.\u001b[39;00m\n\u001b[0;32m--> 449\u001b[0m             six\u001b[39m.\u001b[39;49mraise_from(e, \u001b[39mNone\u001b[39;49;00m)\n\u001b[1;32m    450\u001b[0m \u001b[39mexcept\u001b[39;00m (SocketTimeout, BaseSSLError, SocketError) \u001b[39mas\u001b[39;00m e:\n",
      "File \u001b[0;32m<string>:3\u001b[0m, in \u001b[0;36mraise_from\u001b[0;34m(value, from_value)\u001b[0m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/urllib3/connectionpool.py:444\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    443\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 444\u001b[0m     httplib_response \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49mgetresponse()\n\u001b[1;32m    445\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    446\u001b[0m     \u001b[39m# Remove the TypeError from the exception chain in\u001b[39;00m\n\u001b[1;32m    447\u001b[0m     \u001b[39m# Python 3 (including for exceptions like SystemExit).\u001b[39;00m\n\u001b[1;32m    448\u001b[0m     \u001b[39m# Otherwise it looks like a bug in the code.\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/http/client.py:1374\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1373\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1374\u001b[0m     response\u001b[39m.\u001b[39;49mbegin()\n\u001b[1;32m   1375\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mConnectionError\u001b[39;00m:\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/http/client.py:318\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    317\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[0;32m--> 318\u001b[0m     version, status, reason \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_read_status()\n\u001b[1;32m    319\u001b[0m     \u001b[39mif\u001b[39;00m status \u001b[39m!=\u001b[39m CONTINUE:\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/http/client.py:279\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_read_status\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 279\u001b[0m     line \u001b[39m=\u001b[39m \u001b[39mstr\u001b[39m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mfp\u001b[39m.\u001b[39;49mreadline(_MAXLINE \u001b[39m+\u001b[39;49m \u001b[39m1\u001b[39;49m), \u001b[39m\"\u001b[39m\u001b[39miso-8859-1\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    280\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(line) \u001b[39m>\u001b[39m _MAXLINE:\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/socket.py:705\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    704\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 705\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_sock\u001b[39m.\u001b[39;49mrecv_into(b)\n\u001b[1;32m    706\u001b[0m \u001b[39mexcept\u001b[39;00m timeout:\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/ssl.py:1274\u001b[0m, in \u001b[0;36mSSLSocket.recv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1271\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[1;32m   1272\u001b[0m           \u001b[39m\"\u001b[39m\u001b[39mnon-zero flags not allowed in calls to recv_into() on \u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m\"\u001b[39m \u001b[39m%\u001b[39m\n\u001b[1;32m   1273\u001b[0m           \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m\u001b[39m__class__\u001b[39m)\n\u001b[0;32m-> 1274\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mread(nbytes, buffer)\n\u001b[1;32m   1275\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/ssl.py:1130\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1129\u001b[0m \u001b[39mif\u001b[39;00m buffer \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m-> 1130\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_sslobj\u001b[39m.\u001b[39;49mread(\u001b[39mlen\u001b[39;49m, buffer)\n\u001b[1;32m   1131\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "\u001b[0;31mTimeoutError\u001b[0m: The read operation timed out",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mReadTimeoutError\u001b[0m                          Traceback (most recent call last)",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/requests/adapters.py:489\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    488\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m chunked:\n\u001b[0;32m--> 489\u001b[0m     resp \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49murlopen(\n\u001b[1;32m    490\u001b[0m         method\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mmethod,\n\u001b[1;32m    491\u001b[0m         url\u001b[39m=\u001b[39;49murl,\n\u001b[1;32m    492\u001b[0m         body\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mbody,\n\u001b[1;32m    493\u001b[0m         headers\u001b[39m=\u001b[39;49mrequest\u001b[39m.\u001b[39;49mheaders,\n\u001b[1;32m    494\u001b[0m         redirect\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    495\u001b[0m         assert_same_host\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    496\u001b[0m         preload_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    497\u001b[0m         decode_content\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    498\u001b[0m         retries\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mmax_retries,\n\u001b[1;32m    499\u001b[0m         timeout\u001b[39m=\u001b[39;49mtimeout,\n\u001b[1;32m    500\u001b[0m     )\n\u001b[1;32m    502\u001b[0m \u001b[39m# Send the request.\u001b[39;00m\n\u001b[1;32m    503\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/urllib3/connectionpool.py:787\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    785\u001b[0m     e \u001b[39m=\u001b[39m ProtocolError(\u001b[39m\"\u001b[39m\u001b[39mConnection aborted.\u001b[39m\u001b[39m\"\u001b[39m, e)\n\u001b[0;32m--> 787\u001b[0m retries \u001b[39m=\u001b[39m retries\u001b[39m.\u001b[39;49mincrement(\n\u001b[1;32m    788\u001b[0m     method, url, error\u001b[39m=\u001b[39;49me, _pool\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m, _stacktrace\u001b[39m=\u001b[39;49msys\u001b[39m.\u001b[39;49mexc_info()[\u001b[39m2\u001b[39;49m]\n\u001b[1;32m    789\u001b[0m )\n\u001b[1;32m    790\u001b[0m retries\u001b[39m.\u001b[39msleep()\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/urllib3/util/retry.py:550\u001b[0m, in \u001b[0;36mRetry.increment\u001b[0;34m(self, method, url, response, error, _pool, _stacktrace)\u001b[0m\n\u001b[1;32m    549\u001b[0m \u001b[39mif\u001b[39;00m read \u001b[39mis\u001b[39;00m \u001b[39mFalse\u001b[39;00m \u001b[39mor\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_is_method_retryable(method):\n\u001b[0;32m--> 550\u001b[0m     \u001b[39mraise\u001b[39;00m six\u001b[39m.\u001b[39;49mreraise(\u001b[39mtype\u001b[39;49m(error), error, _stacktrace)\n\u001b[1;32m    551\u001b[0m \u001b[39melif\u001b[39;00m read \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/urllib3/packages/six.py:770\u001b[0m, in \u001b[0;36mreraise\u001b[0;34m(tp, value, tb)\u001b[0m\n\u001b[1;32m    769\u001b[0m         \u001b[39mraise\u001b[39;00m value\u001b[39m.\u001b[39mwith_traceback(tb)\n\u001b[0;32m--> 770\u001b[0m     \u001b[39mraise\u001b[39;00m value\n\u001b[1;32m    771\u001b[0m \u001b[39mfinally\u001b[39;00m:\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/urllib3/connectionpool.py:703\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    702\u001b[0m \u001b[39m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m httplib_response \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_request(\n\u001b[1;32m    704\u001b[0m     conn,\n\u001b[1;32m    705\u001b[0m     method,\n\u001b[1;32m    706\u001b[0m     url,\n\u001b[1;32m    707\u001b[0m     timeout\u001b[39m=\u001b[39;49mtimeout_obj,\n\u001b[1;32m    708\u001b[0m     body\u001b[39m=\u001b[39;49mbody,\n\u001b[1;32m    709\u001b[0m     headers\u001b[39m=\u001b[39;49mheaders,\n\u001b[1;32m    710\u001b[0m     chunked\u001b[39m=\u001b[39;49mchunked,\n\u001b[1;32m    711\u001b[0m )\n\u001b[1;32m    713\u001b[0m \u001b[39m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    714\u001b[0m \u001b[39m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    715\u001b[0m \u001b[39m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    716\u001b[0m \u001b[39m# mess.\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/urllib3/connectionpool.py:451\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    450\u001b[0m \u001b[39mexcept\u001b[39;00m (SocketTimeout, BaseSSLError, SocketError) \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m--> 451\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_raise_timeout(err\u001b[39m=\u001b[39;49me, url\u001b[39m=\u001b[39;49murl, timeout_value\u001b[39m=\u001b[39;49mread_timeout)\n\u001b[1;32m    452\u001b[0m     \u001b[39mraise\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/urllib3/connectionpool.py:340\u001b[0m, in \u001b[0;36mHTTPConnectionPool._raise_timeout\u001b[0;34m(self, err, url, timeout_value)\u001b[0m\n\u001b[1;32m    339\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(err, SocketTimeout):\n\u001b[0;32m--> 340\u001b[0m     \u001b[39mraise\u001b[39;00m ReadTimeoutError(\n\u001b[1;32m    341\u001b[0m         \u001b[39mself\u001b[39m, url, \u001b[39m\"\u001b[39m\u001b[39mRead timed out. (read timeout=\u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m)\u001b[39m\u001b[39m\"\u001b[39m \u001b[39m%\u001b[39m timeout_value\n\u001b[1;32m    342\u001b[0m     )\n\u001b[1;32m    344\u001b[0m \u001b[39m# See the above comment about EAGAIN in Python 3. In Python 2 we have\u001b[39;00m\n\u001b[1;32m    345\u001b[0m \u001b[39m# to specifically catch it and throw the timeout error\u001b[39;00m\n",
      "\u001b[0;31mReadTimeoutError\u001b[0m: HTTPSConnectionPool(host='api.polygon.io', port=443): Read timed out. (read timeout=10)",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mReadTimeout\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 14\u001b[0m\n\u001b[1;32m     11\u001b[0m symbol \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39mRNDR\u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m     12\u001b[0m timeframe \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39m30m\u001b[39m\u001b[39m'\u001b[39m\n\u001b[0;32m---> 14\u001b[0m data \u001b[39m=\u001b[39m get_data_with_features_and_labels(\n\u001b[1;32m     15\u001b[0m     symbol\u001b[39m=\u001b[39;49msymbol,\n\u001b[1;32m     16\u001b[0m     timeframes\u001b[39m=\u001b[39;49m[timeframe],\n\u001b[1;32m     17\u001b[0m     separate\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m     18\u001b[0m     label_args\u001b[39m=\u001b[39;49m(\u001b[39m0.06\u001b[39;49m, \u001b[39m0.0075\u001b[39;49m),\n\u001b[1;32m     19\u001b[0m )\n\u001b[1;32m     21\u001b[0m data[\u001b[39m'\u001b[39m\u001b[39mlabel\u001b[39m\u001b[39m'\u001b[39m][data[\u001b[39m'\u001b[39m\u001b[39mlabel\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m==\u001b[39m \u001b[39m1\u001b[39m] \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m\n\u001b[1;32m     22\u001b[0m data[\u001b[39m'\u001b[39m\u001b[39mlabel\u001b[39m\u001b[39m'\u001b[39m][data[\u001b[39m'\u001b[39m\u001b[39mlabel\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m==\u001b[39m \u001b[39m2\u001b[39m] \u001b[39m=\u001b[39m \u001b[39m1\u001b[39m\n",
      "File \u001b[0;32m~/vctr/vctr/utils/cache.py:39\u001b[0m, in \u001b[0;36mcache_plz.<locals>.decorator.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     36\u001b[0m     \u001b[39mreturn\u001b[39;00m result\n\u001b[1;32m     38\u001b[0m \u001b[39m# Invoke the function and cache the result\u001b[39;00m\n\u001b[0;32m---> 39\u001b[0m result \u001b[39m=\u001b[39m func(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m     40\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m no_cache:\n\u001b[1;32m     41\u001b[0m     \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39m(cache_file, \u001b[39m'\u001b[39m\u001b[39mwb\u001b[39m\u001b[39m'\u001b[39m) \u001b[39mas\u001b[39;00m f:\n",
      "File \u001b[0;32m~/vctr/vctr/data/data_loader.py:182\u001b[0m, in \u001b[0;36mget_data_with_features_and_labels\u001b[0;34m(symbol, timeframes, label_args, start, end, separate, crypto)\u001b[0m\n\u001b[1;32m    179\u001b[0m \u001b[39mif\u001b[39;00m label_args \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m    180\u001b[0m     label_args \u001b[39m=\u001b[39m (\u001b[39m0.05\u001b[39m, \u001b[39m0.01\u001b[39m)\n\u001b[0;32m--> 182\u001b[0m data \u001b[39m=\u001b[39m get_data_with_features(symbol, timeframes, start, end, crypto\u001b[39m=\u001b[39;49mcrypto)\n\u001b[1;32m    183\u001b[0m data \u001b[39m=\u001b[39m label_data_extrema_multi(data, \u001b[39m*\u001b[39mlabel_args)\n\u001b[1;32m    185\u001b[0m \u001b[39m# cat_clf = CatBoostClassifier().load_model('trend_clf.model')\u001b[39;00m\n\u001b[1;32m    186\u001b[0m \u001b[39m# X_clf = StandardScaler().fit_transform(data.copy())\u001b[39;00m\n\u001b[1;32m    187\u001b[0m \u001b[39m# data['trend_clf'] = pd.Series(cat_clf.predict(X_clf), index=data.index)\u001b[39;00m\n\u001b[1;32m    188\u001b[0m \u001b[39m# data = label_data_trends_binary(data)\u001b[39;00m\n",
      "File \u001b[0;32m~/vctr/vctr/utils/cache.py:39\u001b[0m, in \u001b[0;36mcache_plz.<locals>.decorator.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     36\u001b[0m     \u001b[39mreturn\u001b[39;00m result\n\u001b[1;32m     38\u001b[0m \u001b[39m# Invoke the function and cache the result\u001b[39;00m\n\u001b[0;32m---> 39\u001b[0m result \u001b[39m=\u001b[39m func(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m     40\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m no_cache:\n\u001b[1;32m     41\u001b[0m     \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39m(cache_file, \u001b[39m'\u001b[39m\u001b[39mwb\u001b[39m\u001b[39m'\u001b[39m) \u001b[39mas\u001b[39;00m f:\n",
      "File \u001b[0;32m~/vctr/vctr/data/data_loader.py:121\u001b[0m, in \u001b[0;36mget_data_with_features\u001b[0;34m(symbol, timeframes, start, end, crypto)\u001b[0m\n\u001b[1;32m    118\u001b[0m original_timeframe \u001b[39m=\u001b[39m timeframes[\u001b[39m0\u001b[39m]\n\u001b[1;32m    120\u001b[0m \u001b[39mfor\u001b[39;00m timeframe \u001b[39min\u001b[39;00m timeframes:\n\u001b[0;32m--> 121\u001b[0m     data \u001b[39m=\u001b[39m get_data(symbol, timeframe, start, end, crypto\u001b[39m=\u001b[39;49mcrypto)\n\u001b[1;32m    123\u001b[0m     \u001b[39mif\u001b[39;00m data\u001b[39m.\u001b[39mempty:\n\u001b[1;32m    124\u001b[0m         \u001b[39m# Throw an error if no data was fetched\u001b[39;00m\n\u001b[1;32m    125\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mNo data for \u001b[39m\u001b[39m{\u001b[39;00msymbol\u001b[39m}\u001b[39;00m\u001b[39m in \u001b[39m\u001b[39m{\u001b[39;00mtimeframe\u001b[39m}\u001b[39;00m\u001b[39m timeframe\u001b[39m\u001b[39m'\u001b[39m)\n",
      "File \u001b[0;32m~/vctr/vctr/data/data_loader.py:83\u001b[0m, in \u001b[0;36mget_data\u001b[0;34m(symbol, timeframe, start, end, crypto)\u001b[0m\n\u001b[1;32m     81\u001b[0m \u001b[39m# Fetch data for all date range chunks in parallel using a ThreadPoolExecutor.\u001b[39;00m\n\u001b[1;32m     82\u001b[0m \u001b[39mwith\u001b[39;00m ThreadPoolExecutor(max_workers\u001b[39m=\u001b[39mn_workers) \u001b[39mas\u001b[39;00m executor:\n\u001b[0;32m---> 83\u001b[0m     data_frames \u001b[39m=\u001b[39m \u001b[39mlist\u001b[39;49m(executor\u001b[39m.\u001b[39;49mmap(fetch_data_chunk_partial, date_chunks))\n\u001b[1;32m     85\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m data_frames \u001b[39mor\u001b[39;00m \u001b[39mall\u001b[39m(df\u001b[39m.\u001b[39mempty \u001b[39mfor\u001b[39;00m df \u001b[39min\u001b[39;00m data_frames):\n\u001b[1;32m     86\u001b[0m     \u001b[39mreturn\u001b[39;00m pd\u001b[39m.\u001b[39mDataFrame()\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/concurrent/futures/_base.py:621\u001b[0m, in \u001b[0;36mExecutor.map.<locals>.result_iterator\u001b[0;34m()\u001b[0m\n\u001b[1;32m    618\u001b[0m \u001b[39mwhile\u001b[39;00m fs:\n\u001b[1;32m    619\u001b[0m     \u001b[39m# Careful not to keep a reference to the popped future\u001b[39;00m\n\u001b[1;32m    620\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m--> 621\u001b[0m         \u001b[39myield\u001b[39;00m _result_or_cancel(fs\u001b[39m.\u001b[39;49mpop())\n\u001b[1;32m    622\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    623\u001b[0m         \u001b[39myield\u001b[39;00m _result_or_cancel(fs\u001b[39m.\u001b[39mpop(), end_time \u001b[39m-\u001b[39m time\u001b[39m.\u001b[39mmonotonic())\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/concurrent/futures/_base.py:319\u001b[0m, in \u001b[0;36m_result_or_cancel\u001b[0;34m(***failed resolving arguments***)\u001b[0m\n\u001b[1;32m    317\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    318\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 319\u001b[0m         \u001b[39mreturn\u001b[39;00m fut\u001b[39m.\u001b[39;49mresult(timeout)\n\u001b[1;32m    320\u001b[0m     \u001b[39mfinally\u001b[39;00m:\n\u001b[1;32m    321\u001b[0m         fut\u001b[39m.\u001b[39mcancel()\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/concurrent/futures/_base.py:458\u001b[0m, in \u001b[0;36mFuture.result\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    456\u001b[0m     \u001b[39mraise\u001b[39;00m CancelledError()\n\u001b[1;32m    457\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_state \u001b[39m==\u001b[39m FINISHED:\n\u001b[0;32m--> 458\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m__get_result()\n\u001b[1;32m    459\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    460\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mTimeoutError\u001b[39;00m()\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/concurrent/futures/_base.py:403\u001b[0m, in \u001b[0;36mFuture.__get_result\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    401\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_exception:\n\u001b[1;32m    402\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 403\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_exception\n\u001b[1;32m    404\u001b[0m     \u001b[39mfinally\u001b[39;00m:\n\u001b[1;32m    405\u001b[0m         \u001b[39m# Break a reference cycle with the exception in self._exception\u001b[39;00m\n\u001b[1;32m    406\u001b[0m         \u001b[39mself\u001b[39m \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/concurrent/futures/thread.py:58\u001b[0m, in \u001b[0;36m_WorkItem.run\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     55\u001b[0m     \u001b[39mreturn\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m---> 58\u001b[0m     result \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mfn(\u001b[39m*\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mkwargs)\n\u001b[1;32m     59\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m exc:\n\u001b[1;32m     60\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfuture\u001b[39m.\u001b[39mset_exception(exc)\n",
      "File \u001b[0;32m~/vctr/vctr/data/data_loader.py:70\u001b[0m, in \u001b[0;36mget_data.<locals>.fetch_data_chunk\u001b[0;34m(symbol, timeframe, date_range)\u001b[0m\n\u001b[1;32m     68\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mfetch_data_chunk\u001b[39m(symbol: \u001b[39mstr\u001b[39m, timeframe: \u001b[39mstr\u001b[39m, date_range: Tuple[datetime, datetime]) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m pd\u001b[39m.\u001b[39mDataFrame:\n\u001b[1;32m     69\u001b[0m     start, end \u001b[39m=\u001b[39m date_range\n\u001b[0;32m---> 70\u001b[0m     \u001b[39mreturn\u001b[39;00m fetch_polygon_aggs(\n\u001b[1;32m     71\u001b[0m         symbol,\n\u001b[1;32m     72\u001b[0m         timeframe\u001b[39m=\u001b[39;49mtimeframe,\n\u001b[1;32m     73\u001b[0m         start\u001b[39m=\u001b[39;49mstart,\n\u001b[1;32m     74\u001b[0m         end\u001b[39m=\u001b[39;49mend,\n\u001b[1;32m     75\u001b[0m         crypto\u001b[39m=\u001b[39;49mcrypto,\n\u001b[1;32m     76\u001b[0m     )\n",
      "File \u001b[0;32m~/vctr/vctr/data/data_loader.py:212\u001b[0m, in \u001b[0;36mfetch_polygon_aggs\u001b[0;34m(symbol, timeframe, start, end, crypto)\u001b[0m\n\u001b[1;32m    209\u001b[0m timespan \u001b[39m=\u001b[39m \u001b[39mdict\u001b[39m(\u001b[39mzip\u001b[39m(\u001b[39m'\u001b[39m\u001b[39mmhd\u001b[39m\u001b[39m'\u001b[39m, [\u001b[39m'\u001b[39m\u001b[39mminute\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mhour\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mday\u001b[39m\u001b[39m'\u001b[39m]))[timespan]\n\u001b[1;32m    211\u001b[0m \u001b[39mif\u001b[39;00m crypto:\n\u001b[0;32m--> 212\u001b[0m     ohlcv \u001b[39m=\u001b[39m crypto_client\u001b[39m.\u001b[39;49mget_full_range_aggregate_bars(\n\u001b[1;32m    213\u001b[0m         symbol,\n\u001b[1;32m    214\u001b[0m         start,\n\u001b[1;32m    215\u001b[0m         end,\n\u001b[1;32m    216\u001b[0m         multiplier\u001b[39m=\u001b[39;49mmultiplier,\n\u001b[1;32m    217\u001b[0m         timespan\u001b[39m=\u001b[39;49mtimespan,\n\u001b[1;32m    218\u001b[0m         adjusted\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    219\u001b[0m         warnings\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    220\u001b[0m         run_parallel\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m    221\u001b[0m         high_volatility\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m,\n\u001b[1;32m    222\u001b[0m     )\n\u001b[1;32m    223\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    224\u001b[0m     ohlcv \u001b[39m=\u001b[39m stocks_client\u001b[39m.\u001b[39mget_aggregate_bars(\n\u001b[1;32m    225\u001b[0m         symbol,\n\u001b[1;32m    226\u001b[0m         start,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    234\u001b[0m         full_range\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m,\n\u001b[1;32m    235\u001b[0m     )\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/polygon/crypto/crypto_api.py:348\u001b[0m, in \u001b[0;36mSyncCryptoClient.get_full_range_aggregate_bars\u001b[0;34m(self, symbol, from_date, to_date, multiplier, timespan, adjusted, sort, run_parallel, max_concurrent_workers, warnings, high_volatility)\u001b[0m\n\u001b[1;32m    346\u001b[0m \u001b[39m# Sequential Run\u001b[39;00m\n\u001b[1;32m    347\u001b[0m time_chunks \u001b[39m=\u001b[39m [from_date, to_date]\n\u001b[0;32m--> 348\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mget_full_range_aggregates(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mget_aggregate_bars, symbol, time_chunks, run_parallel,\n\u001b[1;32m    349\u001b[0m                                       max_concurrent_workers, warnings, adjusted\u001b[39m=\u001b[39;49madjusted,\n\u001b[1;32m    350\u001b[0m                                       multiplier\u001b[39m=\u001b[39;49mmultiplier, sort\u001b[39m=\u001b[39;49msort, limit\u001b[39m=\u001b[39;49m\u001b[39m50_000\u001b[39;49m,\n\u001b[1;32m    351\u001b[0m                                       timespan\u001b[39m=\u001b[39;49mtimespan)\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/polygon/base_client.py:528\u001b[0m, in \u001b[0;36mBaseClient.get_full_range_aggregates\u001b[0;34m(self, fn, symbol, time_chunks, run_parallel, max_concurrent_workers, warnings, adjusted, sort, limit, multiplier, timespan)\u001b[0m\n\u001b[1;32m    522\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m'\u001b[39m\u001b[39mInvalid timespan. Use a correct enum or a correct value. See \u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m    523\u001b[0m                      \u001b[39m'\u001b[39m\u001b[39mhttps://polygon.readthedocs.io/en/latest/Library-Interface-Documentation.html#polygon\u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m    524\u001b[0m                      \u001b[39m'\u001b[39m\u001b[39m.enums.Timespan\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m    526\u001b[0m \u001b[39mif\u001b[39;00m (\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mnormalize_datetime(end_dt, \u001b[39m'\u001b[39m\u001b[39mdatetime\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mend\u001b[39m\u001b[39m'\u001b[39m) \u001b[39m-\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mnormalize_datetime(\n\u001b[1;32m    527\u001b[0m         first_entry, \u001b[39m'\u001b[39m\u001b[39mdatetime\u001b[39m\u001b[39m'\u001b[39m))\u001b[39m.\u001b[39mdays \u001b[39m<\u001b[39m\u001b[39m=\u001b[39m delta\u001b[39m.\u001b[39mdays:\n\u001b[0;32m--> 528\u001b[0m     res \u001b[39m=\u001b[39m fn(symbol, current_dt, end_dt, adjusted\u001b[39m=\u001b[39;49madjusted, sort\u001b[39m=\u001b[39;49msort, limit\u001b[39m=\u001b[39;49m\u001b[39m500000\u001b[39;49m,\n\u001b[1;32m    529\u001b[0m              multiplier\u001b[39m=\u001b[39;49mmultiplier, timespan\u001b[39m=\u001b[39;49mtimespan, full_range\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m)\n\u001b[1;32m    530\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    531\u001b[0m         \u001b[39mreturn\u001b[39;00m res[\u001b[39m'\u001b[39m\u001b[39mresults\u001b[39m\u001b[39m'\u001b[39m]\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/polygon/crypto/crypto_api.py:282\u001b[0m, in \u001b[0;36mSyncCryptoClient.get_aggregate_bars\u001b[0;34m(self, symbol, from_date, to_date, multiplier, timespan, adjusted, sort, limit, full_range, run_parallel, max_concurrent_workers, warnings, high_volatility, raw_response)\u001b[0m\n\u001b[1;32m    275\u001b[0m _path \u001b[39m=\u001b[39m \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m/v2/aggs/ticker/\u001b[39m\u001b[39m{\u001b[39;00mensure_prefix(symbol)\u001b[39m.\u001b[39mupper()\u001b[39m}\u001b[39;00m\u001b[39m/range/\u001b[39m\u001b[39m{\u001b[39;00mmultiplier\u001b[39m}\u001b[39;00m\u001b[39m/\u001b[39m\u001b[39m{\u001b[39;00mtimespan\u001b[39m}\u001b[39;00m\u001b[39m/\u001b[39m\u001b[39m{\u001b[39;00mfrom_date\u001b[39m}\u001b[39;00m\u001b[39m/\u001b[39m\u001b[39m'\u001b[39m \\\n\u001b[1;32m    276\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mto_date\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m\n\u001b[1;32m    278\u001b[0m _data \u001b[39m=\u001b[39m {\u001b[39m'\u001b[39m\u001b[39madjusted\u001b[39m\u001b[39m'\u001b[39m: \u001b[39m'\u001b[39m\u001b[39mtrue\u001b[39m\u001b[39m'\u001b[39m \u001b[39mif\u001b[39;00m adjusted \u001b[39melse\u001b[39;00m \u001b[39m'\u001b[39m\u001b[39mfalse\u001b[39m\u001b[39m'\u001b[39m,\n\u001b[1;32m    279\u001b[0m          \u001b[39m'\u001b[39m\u001b[39msort\u001b[39m\u001b[39m'\u001b[39m: sort,\n\u001b[1;32m    280\u001b[0m          \u001b[39m'\u001b[39m\u001b[39mlimit\u001b[39m\u001b[39m'\u001b[39m: limit}\n\u001b[0;32m--> 282\u001b[0m _res \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_get_response(_path, params\u001b[39m=\u001b[39;49m_data)\n\u001b[1;32m    284\u001b[0m \u001b[39mif\u001b[39;00m raw_response:\n\u001b[1;32m    285\u001b[0m     \u001b[39mreturn\u001b[39;00m _res\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/polygon/base_client.py:267\u001b[0m, in \u001b[0;36mBaseClient._get_response\u001b[0;34m(self, path, params, raw_response)\u001b[0m\n\u001b[1;32m    256\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_get_response\u001b[39m(\u001b[39mself\u001b[39m, path: \u001b[39mstr\u001b[39m, params: \u001b[39mdict\u001b[39m \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    257\u001b[0m                   raw_response: \u001b[39mbool\u001b[39m \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Union[Response, \u001b[39mdict\u001b[39m]:\n\u001b[1;32m    258\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    259\u001b[0m \u001b[39m    Get response on a path. Meant to be used internally but can be used if you know what you're doing\u001b[39;00m\n\u001b[1;32m    260\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    265\u001b[0m \u001b[39m    :return: A Response object by default. Make ``raw_response=False`` to get JSON decoded Dictionary\u001b[39;00m\n\u001b[1;32m    266\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 267\u001b[0m     _res \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msession\u001b[39m.\u001b[39;49mrequest(\u001b[39m'\u001b[39;49m\u001b[39mGET\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mBASE \u001b[39m+\u001b[39;49m path, params\u001b[39m=\u001b[39;49mparams, timeout\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mtime_out_conf)\n\u001b[1;32m    269\u001b[0m     \u001b[39mif\u001b[39;00m raw_response:\n\u001b[1;32m    270\u001b[0m         \u001b[39mreturn\u001b[39;00m _res\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/requests/sessions.py:587\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    582\u001b[0m send_kwargs \u001b[39m=\u001b[39m {\n\u001b[1;32m    583\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mtimeout\u001b[39m\u001b[39m\"\u001b[39m: timeout,\n\u001b[1;32m    584\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mallow_redirects\u001b[39m\u001b[39m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    585\u001b[0m }\n\u001b[1;32m    586\u001b[0m send_kwargs\u001b[39m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 587\u001b[0m resp \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49msend(prep, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49msend_kwargs)\n\u001b[1;32m    589\u001b[0m \u001b[39mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/requests/sessions.py:701\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    698\u001b[0m start \u001b[39m=\u001b[39m preferred_clock()\n\u001b[1;32m    700\u001b[0m \u001b[39m# Send the request\u001b[39;00m\n\u001b[0;32m--> 701\u001b[0m r \u001b[39m=\u001b[39m adapter\u001b[39m.\u001b[39;49msend(request, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    703\u001b[0m \u001b[39m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    704\u001b[0m elapsed \u001b[39m=\u001b[39m preferred_clock() \u001b[39m-\u001b[39m start\n",
      "File \u001b[0;32m/opt/homebrew/Caskroom/miniforge/base/envs/v2/lib/python3.10/site-packages/requests/adapters.py:578\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    576\u001b[0m     \u001b[39mraise\u001b[39;00m SSLError(e, request\u001b[39m=\u001b[39mrequest)\n\u001b[1;32m    577\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39misinstance\u001b[39m(e, ReadTimeoutError):\n\u001b[0;32m--> 578\u001b[0m     \u001b[39mraise\u001b[39;00m ReadTimeout(e, request\u001b[39m=\u001b[39mrequest)\n\u001b[1;32m    579\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39misinstance\u001b[39m(e, _InvalidHeader):\n\u001b[1;32m    580\u001b[0m     \u001b[39mraise\u001b[39;00m InvalidHeader(e, request\u001b[39m=\u001b[39mrequest)\n",
      "\u001b[0;31mReadTimeout\u001b[0m: HTTPSConnectionPool(host='api.polygon.io', port=443): Read timed out. (read timeout=10)"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import classification_report, balanced_accuracy_score\n",
    "from catboost import CatBoostClassifier\n",
    "from sklearn.utils.class_weight import compute_class_weight\n",
    "\n",
    "from vctr.data.data_loader import get_data_with_features_and_labels\n",
    "\n",
    "symbol = 'RNDR'\n",
    "timeframe = '30m'\n",
    "\n",
    "data = get_data_with_features_and_labels(\n",
    "    symbol=symbol,\n",
    "    timeframes=[timeframe],\n",
    "    separate=False,\n",
    "    label_args=(0.06, 0.0075),\n",
    ")\n",
    "\n",
    "data['label'][data['label'] == 1] = 0\n",
    "data['label'][data['label'] == 2] = 1\n",
    "\n",
    "# data['close'].vbt.overlay_with_heatmap(data['label']).show()\n",
    "\n",
    "X, y = data.drop(['label'], axis=1), data['label']\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "X_train, X_test = StandardScaler().fit_transform(X_train), StandardScaler().fit_transform(X_test)\n",
    "\n",
    "# Calculate class weights\n",
    "class_weights = compute_class_weight('balanced', classes=np.unique(y_train), y=y_train)\n",
    "class_weights_dict = dict(zip(np.unique(y_train), class_weights))\n",
    "\n",
    "# CatBoost with class weights\n",
    "catboost_clf = CatBoostClassifier(\n",
    "    iterations=750, learning_rate=0.1, class_weights=class_weights_dict, verbose=False\n",
    ")\n",
    "catboost_clf.fit(X_train, y_train)\n",
    "catboost_preds = catboost_clf.predict(X_test)\n",
    "\n",
    "# Evaluate classifiers using balanced accuracy\n",
    "print('CatBoost balanced accuracy: ', balanced_accuracy_score(y_test, catboost_preds))\n",
    "print('\\nCatBoost classification report:\\n', classification_report(y_test, catboost_preds))\n",
    "\n",
    "catboost_clf.save_model('sell_clf.model')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from catboost import CatBoostClassifier\n",
    "import vectorbtpro as vbt\n",
    "\n",
    "from vctr.data.data_loader import get_data_with_features_and_labels\n",
    "from vctr.utils.pf_stats import get_plot_and_pf, print_stats\n",
    "import faulthandler\n",
    "\n",
    "faulthandler.enable()\n",
    "\n",
    "buy_clf = CatBoostClassifier().load_model('buy_clf.model')\n",
    "sell_clf = CatBoostClassifier().load_model('sell_clf.model')\n",
    "\n",
    "symbol = 'ADA'\n",
    "timeframe = '15m'\n",
    "\n",
    "data = get_data_with_features_and_labels(\n",
    "    symbol=symbol,\n",
    "    timeframe=timeframe,\n",
    "    start='2023-01-01',\n",
    "    separate=False,\n",
    "    label_args=(0.03, 0.005),\n",
    ")\n",
    "\n",
    "X, y = data.drop(['label'], axis=1), data['label']\n",
    "X = StandardScaler().fit_transform(X)\n",
    "\n",
    "buy_preds = buy_clf.predict(X)\n",
    "sell_preds = sell_clf.predict(X)\n",
    "\n",
    "data = data[-len(sell_preds) :].copy()\n",
    "data['buy_pred'] = pd.Series(buy_preds, index=data.index)\n",
    "data['sell_pred'] = pd.Series(sell_preds, index=data.index)\n",
    "\n",
    "pf = vbt.Portfolio.from_signals(\n",
    "    data['close'],\n",
    "    data['buy_pred'] == 1,\n",
    "    data['sell_pred'] == 1,\n",
    "    freq='15T',\n",
    "    init_cash=10000,\n",
    "    fees=0.0006,\n",
    "    slippage=0.001,\n",
    "    log=True,\n",
    ")\n",
    "# fig = pf.trades.plot()\n",
    "\n",
    "# axis_args = dict(showgrid=True, gridwidth=0.5, gridcolor='rgba(255, 255, 255, 0.05)', griddash='dot')\n",
    "# fig.update_xaxes(**axis_args)\n",
    "# fig.update_yaxes(**axis_args)\n",
    "# fig.update_layout(dict(width=1050, height=600))\n",
    "\n",
    "# fig.data[0].line.width = 1\n",
    "# fig.data[0].opacity = 0.75\n",
    "# fig.show()\n",
    "print_stats(pf)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "v2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
